Random failing IResource.testDelete

The test `org.eclipse.core.tests.resources.IResourceTest.testDelete` fails randomly.
See https://github.com/eclipse-platform/eclipse.platform/actions/runs/4742853133/jobs/8421637788?pr=415#step:7:17743
```
java.lang.AssertionError: invocation 73 should fail, but it doesn't [false,null,F/openProject/1/2/3]
	at org.junit.Assert.fail(Assert.java:89)
	at org.eclipse.core.tests.resources.TestPerformer.performTestRecursiveLoop(TestPerformer.java:81)
	at org.eclipse.core.tests.resources.TestPerformer.performTestRecursiveLoop(TestPerformer.java:110)
	at org.eclipse.core.tests.resources.TestPerformer.performTestRecursiveLoop(TestPerformer.java:110)
	at org.eclipse.core.tests.resources.TestPerformer.performTest(TestPerformer.java:56)
	at org.eclipse.core.tests.resources.IResourceTest.testDelete(IResourceTest.java:1132)
```

Failure occurred on MacOS.

There is another recent failure of a test using the `TestPerformer`: `org.eclipse.core.tests.resources.LinkedResourceTest.testCopyFolder`.
Maybe there is a relation by using that performer class.

See https://github.com/eclipse-platform/eclipse.platform/actions/runs/4722513909/jobs/8377196657#step:7:18606
```
Caused by: org.eclipse.core.internal.resources.ResourceException: Resource '/ExistingProject/existingFolderInExistingProject/nonExistingFolderInExistingFolder' already exists.
	at org.eclipse.core.internal.resources.Resource.checkDoesNotExist(Resource.java:309)
	at org.eclipse.core.internal.resources.Resource.checkDoesNotExist(Resource.java:296)
	at org.eclipse.core.internal.resources.Resource.checkCopyRequirements(Resource.java:248)
	at org.eclipse.core.internal.resources.Resource.assertCopyRequirements(Resource.java:149)
	at org.eclipse.core.internal.resources.Resource.copy(Resource.java:538)
	at org.eclipse.core.tests.resources.LinkedResourceTest$2.invokeMethod(LinkedResourceTest.java:474)
```
Failure occurred on Windows. So if there is a relation, the problems are independent from OS / file system.

By looking at the test classes `LinkedResourceTest` and `IResourceTest` I can tell, they are almost bound to have problems: both classes share resources (`IResource`s / `IFile`s / `IFolder`s / ...) among several (of their own) test methods and those resources are moved around, copied, deleted, etc.

Shared resources in tests are always a problem because every test method needs to:
a) Thoroughly clean up after finishing (_i.e._ if the operating system blocks a file/folder then the next test using the resource will fail) and
b) It keeps the tests from parallel execution

There is no easy way to reproduce the error, but you could try the following:
* Debug a test
* Stop right before cleaning up (set a breakpoint in the first line of `org.eclipse.core.tests.resources.LinkedResourceTest.doCleanup()`)
* Lock a resource: in Windows you could open the command-line and `cd` into the location of `existingFolderInExistingProject` (it's inside the WS)
* Let the clean-up continue (it should fail)
* Start the test again
* As long as the folder is still locked, successive "clean-ups" should fail and any call to `IResource::move` with an existing folder as the destination (a folder that has not been cleaned-up) should fail too.

Also, I see that the second error you posted refers to a failure in `LinkedResourceWithPathVariableTest.testCopyFolder` which is actually an unnecessary execution of `testCopyFolder` in the super class `LinkedResourceTest`. 

```
org.eclipse.core.tests.resources.LinkedResourceWithPathVariableTest.testCopyFolder  Time elapsed: 0.066 s  <<< FAILURE!
java.lang.AssertionError: invocation 16 failed with org.eclipse.core.internal.resources.ResourceException: Resource '/ExistingProject/existingFolderInExistingProject/nonExistingFolderInExistingFolder' already exists.
	at org.eclipse.core.tests.resources.TestPerformer.performTestRecursiveLoop(TestPerformer.java:97)
	at org.eclipse.core.tests.resources.TestPerformer.performTestRecursiveLoop(TestPerformer.java:110)
	at org.eclipse.core.tests.resources.TestPerformer.performTestRecursiveLoop(TestPerformer.java:110)
	at org.eclipse.core.tests.resources.TestPerformer.performTestRecursiveLoop(TestPerformer.java:110)
	at org.eclipse.core.tests.resources.TestPerformer.performTest(TestPerformer.java:56)
	at org.eclipse.core.tests.resources.LinkedResourceTest.testCopyFolder(LinkedResourceTest.java:552)
```

Extending classes with test methods just to be able to use some of their methods/fields is also a bad practice because it leads to running the same test at least twice.

Hope this helps!

--- a/resources/tests/org.eclipse.core.tests.resources/src/org/eclipse/core/tests/internal/builders/ParallelBuildChainTest.java
+++ b/resources/tests/org.eclipse.core.tests.resources/src/org/eclipse/core/tests/internal/builders/ParallelBuildChainTest.java
@@ -19,6 +19,7 @@ import org.eclipse.core.resources.*;
 import org.eclipse.core.runtime.*;
 import org.eclipse.core.runtime.jobs.Job;
 import org.eclipse.core.runtime.jobs.JobGroup;
+import org.eclipse.core.tests.harness.TestBarrier2;
 import org.eclipse.core.tests.internal.builders.TimerBuilder.RuleType;
 import org.junit.Assert;
 import org.junit.Test;
@@ -237,12 +238,14 @@ public class ParallelBuildChainTest extends AbstractBuilderTest {
 
 	@Test
 	public void testWorkspaceParrallelBuildCurrentProject() throws CoreException, OperationCanceledException, InterruptedException {
+		TestBarrier2 waitForRunningJobBarrier = new TestBarrier2();
 		setTimerBuilderSchedulingRuleForAllProjects(RuleType.CURRENT_PROJECT, getMonitor());
 		long duration = System.currentTimeMillis();
 		Job job = new Job("Workspace Build") {
 			@Override
 			protected IStatus run(IProgressMonitor monitor) {
 				try {
+					waitForRunningJobBarrier.setStatus(TestBarrier2.STATUS_RUNNING);
 					getWorkspace().build(IncrementalProjectBuilder.INCREMENTAL_BUILD, getMonitor());
 					return Status.OK_STATUS;
 				} catch (CoreException e) {
@@ -250,8 +253,10 @@ public class ParallelBuildChainTest extends AbstractBuilderTest {
 				}
 			}
 		};
+
 		job.schedule();
-		Assert.assertTrue("Timeout, most likely a deadlock", job.join(5000, getMonitor()));
+		waitForRunningJobBarrier.waitForStatus(TestBarrier2.STATUS_RUNNING);
+		Assert.assertTrue("Timeout, most likely a deadlock", job.join(20000, getMonitor()));
 		duration = System.currentTimeMillis() - duration;
 		assertEquals(getWorkspace().getRoot().getProjects().length, TimerBuilder.getTotalBuilds());
 		assertTrue(TimerBuilder.getMaxSimultaneousBuilds() > 1);

--- a/resources/tests/org.eclipse.core.tests.resources/src/org/eclipse/core/tests/resources/ResourceTest.java
+++ b/resources/tests/org.eclipse.core.tests.resources/src/org/eclipse/core/tests/resources/ResourceTest.java
@@ -416,6 +416,8 @@ public abstract class ResourceTest extends CoreTest {
 	}
 
 	protected void cleanup() throws CoreException {
+		// Wait for any build job that may still be executed
+		waitForBuild();
 		final IFileStore[] toDelete = storesToDelete.toArray(new IFileStore[0]);
 		storesToDelete.clear();
 		getWorkspace().run((IWorkspaceRunnable) monitor -> {
@@ -426,7 +428,7 @@ public abstract class ResourceTest extends CoreTest {
 			}
 		}, null);
 		getWorkspace().save(true, null);
-		//don't leak builder jobs, since they may affect subsequent tests
+		// don't leak builder jobs, since they may affect subsequent tests
 		waitForBuild();
 		assertWorkspaceFolderEmpty();
 	}

